### PR TITLE
Run Llama2 with torch.compile on Gaudi2

### PR DESCRIPTION
**Pull request by kausikmaiti**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by vivekgoe** [comment]
@kausikmaiti looks good to me. @regisss please review and help merge this if it looks ok to you. Thanks.

**Review by vivekgoe — APPROVED** [review]

**Review by libinta — COMMENTED** [review]
@kausikmaiti I have following questions
1. what's the default cmd for llama2? torch compile or use_lazy_mode? if the default is torch.compile, please also update the readme
2. what's the performance comparison with use_lazy_mode and use_hpu_graph?
3. any depenency to the to-be-released docker?

File: examples/text-generation/utils.py
Comment (regisss) [inline examples/text-generation/utils.py:155]: Is "model" a new attribute of `model` or does it already exist?

**Comment by kausikmaiti** [comment]
> @kausikmaiti I have following questions
> 
> 1. what's the default cmd for llama2? torch compile or use_lazy_mode? if the default is torch.compile, please also update the readme
> 2. what's the performance comparison with use_lazy_mode and use_hpu_graph?
> 3. any depenency to the to-be-released docker?

@libinta, Please find my answers below.
1. Here, I am just adding support for torch.compile. But, default mode is still unchanged. So, we don't need to update readme here.
2. This is the very first step of (functional) enablement. Performance data is not yet available.
3. Yes. It has dependency on the content of release 1.14.

File: examples/text-generation/utils.py
Comment (kausikmaiti) [reply examples/text-generation/utils.py:155]: @regisss, It already exists.

File: examples/text-generation/run_generation.py
Comment (bhargaveede) [inline examples/text-generation/run_generation.py:257]: We could avoid this by using args.torch_compile directly in 312 and 492 with lazy_mode = not args.torch_compile.

**Comment by regisss** [comment]
> Yes. It has dependency on the content of release 1.14.

So we should not merge it before 1.14 is released @libinta right?

File: examples/text-generation/run_generation.py
Comment (kausikmaiti) [reply examples/text-generation/run_generation.py:257]: Yes, that can be done of course. But, I wanted to have the decision making about use_lazy_mode in one place for all. A secondary reason is that I plan to include additional checks in future PR to control use_lazy_mode.

File: examples/text-generation/utils.py
Comment (bhargaveede) [inline examples/text-generation/utils.py:180]: @kausikmaiti  Can we add model specific check as generation using torch.compile isn't verified on models

Comment (kausikmaiti) [reply examples/text-generation/utils.py:180]: Added model specific check in separate commit. Please review.

**Comment by regisss** [comment]
@kausikmaiti Let's also add a test to: https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py

You can define a new test at the end of this file:
```py
@pytest.mark.parametrize("model_name, baseline", MODELS_TO_TEST["torch_compile"])
def test_text_generation_torch_compile(model_name: str, baseline: float, token: str):
    _test_text_generation(model_name, baseline, token, torch_compile=True)
```
and adding a new `torch_compile` entry in the dict here: https://github.com/huggingface/optimum-habana/blob/979c13247027fa9c8b9cba5137c4a17e27615589/tests/test_text_generation_example.py#L15

File: tests/test_text_generation_example.py
Comment (kausikmaiti) [reply tests/test_text_generation_example.py:1]: As you mentioned offline, WORLD_SIZE setting does not matter, as I'm not using deepspeed / gaudi_spawn.py script.
Also as per my observation, if I don't set WORLD_SIZE=0, due to the logic like "use_deepspeed = args.world_size > 0", setup_distributed_model() gets called and the test fails at very early stage while importing deepspeed. This is not the expectation.

**Review by regisss — APPROVED** [review]

**Pull request merged**
